tests/memoryview/memoryview_annotation_typing.py

```diff
@@ -37,6 +41,34 @@ def two_dim(a: cython.double[:,:]):
     a[0,0] *= 3
     return a[0,0], a[0,1], a.ndim
 
+if cython.compiled:
+    # Tests are run both as compiled and not compiled. slice_none()
+    # output differs for compiled and not compiled. Hence, we need
+    # to ensure that it is run only when it is compiled.
+
+    def slice_none(m: cython.double[:]):
+        """
+        >>> slice_none(None)
+        Traceback (most recent call last):
+          ...
+        TypeError: Argument 'm' must not be None
+        """
+        return m
```

Comment:
This way of writing (or excluding) the test makes it easy to end up with a test that never runs, just by accident. Let's always run it and make sure that we get the expected result in both cases.
```suggestion
def slice_none(m: cython.double[:]):
    """
    >>> try: slice_none(None)
    ... except TypeError as exc:
    ...     assert cython.compiled
    ...     if "Argument 'm' must not be None" not in str(exc): raise
    ... else:
    ...     assert not cython.compiled

    >>> try: slice_none(numpy.ones((10,), numpy.float))
    ... except ValueError as exc:
    ...     assert cython.compiled
    ... else:
    ...     assert not cython.compiled
    """
    return m
```

Reply:
Ah, right. Then just assign `cython.compiled` to a global constant in the module and use that in the doctests.

Reply:
The other option is to put the tests in the global module-level  `__doc__` (with a comment). And use that to pick a different set of tests:

```
if cython.compiled:
  __doc__ = ...
else:
  __doc__ = ...
```

That's definitely something I've done before in similar cases

Reply:
> Ah, right. Then just assign `cython.compiled` to a global constant in the module and use that in the doctests.

OK done. Have a look. I hope now it is fine.

ps. I removed the part with `ValueError` because 
a) the code does not raise an `ValueError` but rather value which differs between compiled and not compiled version
b) the test is about Exception related to `None` not about passing value with different type.

Reply:
> I removed the part with `ValueError`

I meant this as a test that the type inside of the `Optional[…]` is correctly interpreted and applied. Otherwise, that's not actually tested anywhere. I would prefer having at least some kind of such a test running. Sorry for not actually testing my proposal upfront.